### PR TITLE
🇻🇳Rxfactor SelectCountryViewModel

### PR DESCRIPTION
**Pull request by khoi**
## What's this?
Rxfactor of country selection screen.

### Relationships
Part of #3352

## Why do we want this?
🔫 mvx

## How is it done?
The VM exposes `IObservable<IEnumerable<SelectableCountryViewModel>>` which get binded to the table.

### Side effects
I don't think there is any.

## Review hints
None.

The implementation is similar to #3600

## :squid: Permissions
🦑

### CODE REVIEW

**Comment by khoi** [comment]
marked as `wip` as tests are failing 🤦🏼‍♂️

File: Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs [hunk 1]
```diff
@@ -1,50 +1,50 @@
...
-using static Toggl.Multivac.Extensions.StringExtensions;
 using Toggl.Foundation.Interactors;
 
 namespace Toggl.Foundation.MvvmCross.ViewModels
 {
     [Preserve(AllMembers = true)]
     public sealed class SelectCountryViewModel : MvxViewModel<long?, long?>
     {
-        private readonly IMvxNavigationService navigationService;
+        public UIAction Close { get; }
```
Comment (heytherewill) [inline Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs]: There's a close extension, so this property is not needed

File: Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs [hunk 2]
```diff
@@ -1,50 +1,50 @@
...
-        private long? selectedCountryId;
+        public InputAction<SelectableCountryViewModel> SelectCountry { get; }
+
+        public IObservable<IEnumerable<SelectableCountryViewModel>> Countries { get; private set; }
 
-        public IMvxAsyncCommand CloseCommand { get; }
+        public InputAction<string> SetFilterText { get; }
 
-        public IMvxAsyncCommand<SelectableCountryViewModel> SelectCountryCommand { get; }
+        private ISubject<string> filterText = new BehaviorSubject<string>(string.Empty);
```
Comment (heytherewill) [inline Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs]: Make this readonly

File: Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs [hunk 3]
```diff
@@ -1,50 +1,50 @@
...
 
-        public MvxObservableCollection<SelectableCountryViewModel> Suggestions { get; }
-            = new MvxObservableCollection<SelectableCountryViewModel>();
+        private readonly IMvxNavigationService navigationService;
 
-        public string Text { get; set; } = "";
+        private long? selectedCountryId;
```
Comment (heytherewill) [inline Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs]: private fields come before props

File: Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs [hunk 4]
```diff
@@ -53,32 +53,30 @@ public override async Task Initialize()
                 allCountries.Insert(0, selectedElement);
             }
 
-            Suggestions.AddRange(
-                allCountries.Select(country => new SelectableCountryViewModel(
-                    country,
-                    country.Id == selectedCountryId)));
+            Countries = filterText
+                .Select(text =>
```
Comment (heytherewill) [inline Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs]: This would read (and perform) better like this:

```csharp
filterText
    .Select(text => text.Trim())
    .DistinctUntilChanged()
    .Select(countriesFromText)
```

File: Toggl.Daneel/ViewControllers/SelectCountryViewController.cs [hunk 5]
```diff
@@ -20,29 +27,41 @@ public override void ViewDidLoad()
         {
             base.ViewDidLoad();
 
-            var source = new CountryTableViewSource(CountriesTableView);
-            CountriesTableView.Source = source;
+            CountriesTableView.SeparatorStyle = UITableViewCellSeparatorStyle.None;
```
Comment (simonrozsival) [inline Toggl.Daneel/ViewControllers/SelectCountryViewController.cs]: Why did you move these lines from the constructor of the view source to the view controller? 🤔 If the problem is that the constructor is called too soon (you are creating it even before the constructor of the VC is called on line 20), then just create the new instance of the view source here in `ViewDidLoad`.

File: Toggl.Daneel/ViewSources/CountryTableViewSource.cs [hunk 6]
```diff
@@ -1,37 +1,34 @@
...
+            => Observable
+                .FromEventPattern<SelectableCountryViewModel>(e => OnItemTapped += e, e => OnItemTapped -= e)
+                .Select(e => e.EventArgs);
+
         private const string cellIdentifier = nameof(CountryViewCell);
+        private const int rowHeight = 48;
 
-        public CountryTableViewSource(UITableView tableView)
-            : base(tableView)
+        public CountryTableViewSource() : base(new ImmutableArray<SelectableCountryViewModel>(), cellIdentifier)
```
Comment (simonrozsival) [inline Toggl.Daneel/ViewSources/CountryTableViewSource.cs]: So much information on one line - let's put `: base(...` on a separate line please! 😄

File: Toggl.Daneel/ViewControllers/SelectCountryViewController.cs (hunk 5 above)
Comment (simonrozsival) [reply Toggl.Daneel/ViewControllers/SelectCountryViewController.cs]: OK

File: Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs [hunk 7]
```diff
@@ -53,32 +50,28 @@ public override async Task Initialize()
...
             }
 
-            Suggestions.AddRange(
-                allCountries.Select(country => new SelectableCountryViewModel(
-                    country,
-                    country.Id == selectedCountryId)));
+            Countries = filterText
+                .Select(text => text.Trim())
+                .DistinctUntilChanged()
+                .Select(trimmedText =>
```
Comment (heytherewill) [inline Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs]: I'd move this into a nicely named method instead of using a closure

Comment (khoi) [reply Toggl.Foundation.MvvmCross/ViewModels/SelectCountryViewModel.cs]: hmm, I tried to extract this out to a local function, but it doesn't look nicer. I prefer the closure version better. How would you name the method?
